### PR TITLE
Auto update md from ur

### PR DESCRIPTION
**Pull request by jamesgorrie**
__WIP__

~~Waiting for merging of #1277 to add the UI - but the Dynamo Side now works.~~

The problem with before was that we were patching a record that potentially didn't exist, which threw an error. 

Since we're retrieving the `usageRights` anyhow to be able to convert them to metadata, we can just retrieve the whole record and then patch them programatically.

### CODE REVIEW

**Comment by jamesgorrie** [comment]
Back in action with the new and improved (read: Working) update method.

**Comment by jamesgorrie** [comment]
Added some tests as this is an automatic step now, so thought it wise

File: kahuna/public/js/usage-rights/usage-rights-editor.js [hunk 1]
```diff
@@ -152,7 +152,8 @@ usageRightsEditor.controller(
             const image = usageRights.image;
             const resource = image.data.userMetadata.data.usageRights;
             return editsService.update(resource, data, image).
-                then(resource => resource.data);
+                then(resource => resource.data).
+                then(() => setMetadataFromUsageRights(image));
```
Comment (theefer) [inline kahuna/public/js/usage-rights/usage-rights-editor.js]: Is anything looking at the return value of this?

File: metadata-editor/app/controllers/EditsController.scala [hunk 2]
```diff
@@ -146,19 +144,32 @@ object EditsController extends Controller with ArgoHelpers with DynamoEdits {
...
+        // TODO: Be more intelligent about this
+        val mergedMetadata = (metadata.credit, metadata.byline) match {
+          case (Some(_), Some(_)) =>
+            originalMetadata.copy(credit = metadata.credit, byline = metadata.byline)
+          case (Some(_), None) =>
+            originalMetadata.copy(credit = metadata.credit)
+          case (None, Some(_)) =>
+            originalMetadata.copy(byline = metadata.byline)
+          case _ => originalMetadata
+        }
```
Comment (theefer) [inline metadata-editor/app/controllers/EditsController.scala]: Can we not use the `Option` to make this simpler?

`credit = metadata.credit orElse originalMetadata.credit` etc ?

File: metadata-editor/app/controllers/EditsController.scala [hunk 3]
```diff
@@ -146,19 +144,32 @@ object EditsController extends Controller with ArgoHelpers with DynamoEdits {
...
+
+        dynamo.jsonAdd(id, "metadata", metadataAsMap(mergedMetadata))
           .map(publish(id))
-          .map(edits => respond(edits.metadata))
+          .map(edits => respond(edits.metadata, uri = Some(metadataUri(id))))
       }
       .getOrElse(Future.failed(EditsValidationError("no-matching-metadata-found", "Couldn't find any matching metadata")))
```
Comment (theefer) [inline metadata-editor/app/controllers/EditsController.scala]: I'm probably missing something but does that return an error every time a UR is saved that is not a photographer?

**Comment by theefer** [comment]
Minor comments, looks :+1: otherwise. Thorough testing etc..

File: metadata-editor/app/controllers/EditsController.scala (hunk 3 above)
Comment (jamesgorrie) [inline metadata-editor/app/controllers/EditsController.scala]: [A 404 yes](https://github.com/guardian/grid/pull/1251#discussion_r39617185
)

File: kahuna/public/js/usage-rights/usage-rights-editor.js (hunk 1 above)
Comment (jamesgorrie) [inline kahuna/public/js/usage-rights/usage-rights-editor.js]: [Yip](https://github.com/guardian/grid/pull/1283/files#diff-81f57b8057236574f9266cb5d4cb54e3R139)

File: metadata-editor/app/controllers/EditsController.scala (hunk 2 above)
Comment (jamesgorrie) [inline metadata-editor/app/controllers/EditsController.scala]: Thanks, I really didn't like that bit of code.

File: kahuna/public/js/usage-rights/usage-rights-editor.js (hunk 1 above)
Comment (theefer) [inline kahuna/public/js/usage-rights/usage-rights-editor.js]: I meant the value the Promise is resolved with. The returned value is now different, but it seems like it's ignored, so should be OK?

Comment (jamesgorrie) [inline kahuna/public/js/usage-rights/usage-rights-editor.js]: Ah yes - no, no-one was listening to the result of the promise.

File: metadata-editor/app/controllers/EditsController.scala (hunk 3 above)
Comment (jamesgorrie) [inline metadata-editor/app/controllers/EditsController.scala]: Saying that -I wasn't handling the 404 - [I've added that now](https://github.com/guardian/grid/commit/ca24f94cbff568deb4f9e8049424767ef33bdc8e).

**Comment by jamesgorrie** [comment]
Added 404 handling - so awaiting new :+1:

File: metadata-editor/app/lib/UsageRightsMetadataMapper.scala [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+
+import com.gu.mediaservice.model._
+
+object UsageRightsMetadataMapper {
+
+  def usageRightsToMetadata(usageRights: UsageRights): Option[ImageMetadata] = {
+    val toImageMetadata: PartialFunction[UsageRights, ImageMetadata] = (ur: UsageRights) => ur match {
+      case u: StaffPhotographer        => ImageMetadata(byline = Some(u.photographer), credit = Some(u.publication))
+      case u: ContractPhotographer     => ImageMetadata(byline = Some(u.photographer), credit = u.publication)
+      case u: CommissionedPhotographer => ImageMetadata(byline = Some(u.photographer), credit = u.publication)
```
Comment (akash1810) [inline metadata-editor/app/lib/UsageRightsMetadataMapper.scala]: Could we use the [Photographer trait](https://github.com/guardian/grid/blob/master/common-lib/src/main/scala/com/gu/mediaservice/model/UsageRights.scala#L89) here?

Comment (jamesgorrie) [inline metadata-editor/app/lib/UsageRightsMetadataMapper.scala]: Nope - need to access to the case class parameters, which are different from Photographer to photographer.

Comment (jamesgorrie) [inline metadata-editor/app/lib/UsageRightsMetadataMapper.scala]: See `publication` is `String` in StaffPhotographer, but `Option[String]` in `ContractPhotographer`.

**Comment by jamesgorrie** [comment]
@theefer - no more 404 - we just return the unmodified metadata entity.

**Comment by theefer** [comment]
:+1:

**Pull request merged**